### PR TITLE
Core/Player: Implement PvP Talents

### PR DESCRIPTION
**Pull request by Golrag**
**Changes proposed:**

-  Implement PvP Talents

**Target branch(es):**
- master

**Issues addressed:** Closes #  (insert issue tracker number)

**Tests performed:**
Players: A B & C

| TestCase | Passed |
| :--- | :---: |
| A starts duel, B & A should have pvp talents enabled | ✔|
| A starts duel and wins. A & B should have their pvp talents disabled in while not world pvp combat | ✔|
| A starts duel and loses. A & B should have their pvp talents disabled in while not world pvp combat | ✔|
| A starts duel with B. Another C, from enemy faction, starts attacking them. Duel ends, A & C should have pvp talents. B also if he is in combat with C. | :x: |
| Swapping A & B in above test cases does not change the result of the test cases.|
| Changing talents while pvp talents are **enabled** should swap the talents and unlearn the previous spell if not on cooldown. | ✔|
| Changing talents while pvp talents are **disabled** should swap the talents and unlearn the previous spell if not on cooldown. | ✔|
| Entering a battleground should enable pvp talents | ✔ |
| Leaving a battleground should disable pvp talents | ✔ |
| Entering wintergrasp while it's not in battle should enable pvp talents | ✔ |
| Leaving wintergrasp should disable pvp talents if not in wpvp | ✔ |
| Entering a wpvp area (gurubashi arena, warrior OH arena) should enable pvp talents | ✔ |
| Leaving a wpvp area while not in wpvp should disable pvp talents | ✔ |
| Leaving a wpvp area while in wpvp combat should keep pvp talents enabled | ✔ |
| Finishing a duel in a wpvp area should keep pvp talents enabled | ✔ |

**Known issues and TODO list:**
- [x] Talents should be activated when duel countdown is started
- [x] Remove that AreaID check in ActivatesPvpTalents if possible. Tol Barad is an exception right now.
- [x] Learn 195710  (Honorable Medallion). Would like to know if you learn it in open world pvp :)
- [x] Duel bug: see description in comment below.
- Fix failing tests in the section above if they are in the current scope of the PR.
- Duel & wpvp bug. CombatTimer gets set to 0 when duel ends, not touching that code as it does not really belong here.

**Notes:**
- It's just a big copy-paste of the current talent system

Character SQL:
```sql
DROP TABLE IF EXISTS `character_pvp_talent`;
CREATE TABLE `character_pvp_talent` (
  `Guid` bigint(20) unsigned NOT NULL,
  `TalentID` mediumint(8) unsigned NOT NULL,
  `TalentGroup` tinyint(3) unsigned NOT NULL DEFAULT '0'
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

Any feedback and help is welcome :)

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: This function is wrong. Being in enemy zone or even capital city does not enable pvp talents.
The only areas in outside world that do are pvp world quest areas and tol barad/wintergrasp (only during battle)

File: src/server/game/Entities/Player/Player.cpp
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: This is wrong - pvp talents remain active for 20 seconds after leaving combat (tied to aura 134735)

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Player/Player.cpp
Comment (Golrag) [inline src/server/game/Entities/Player/Player.cpp]: Baited. Oh well, FFA areas also activates them btw (gurubashi arena, warrior order hall arena)
Will look into it tomorrow

Comment (Golrag) [inline src/server/game/Entities/Player/Player.cpp]: What about adding 134735 to spell_area table ?
It gets applied without a duration when entering pvp areas and instantly removed when leaving them (without combat involved). Also duration of the aura in spellwork says 20 seconds, so when leaving combat the duration should start (I think)

Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: But that will make you get that aura with 20s duration

Comment (Golrag) [inline src/server/game/Entities/Player/Player.cpp]: Yeah it would :(

Comment (Golrag) [inline src/server/game/Entities/Player/Player.cpp]: Is there a function that changes the duration of the spell you want to cast ?

Comment (Warpten) [inline src/server/game/Entities/Player/Player.cpp]: you can do that in an aurascript iirc

Comment (Golrag) [inline src/server/game/Entities/Player/Player.cpp]: Could be, but I would still need something to make it back to 20sec on exit combat :p

Comment (Golrag) [inline src/server/game/Entities/Player/Player.cpp]: I got something (local) for the AuraScript and using the spell_area but I noticed that when we cast the spell outside the areas specified in spell_area table we get SPELL_FAILED_INCORRECT_AREA. So I was thinking about just having a function called EnablesPvpTalents in AreaTableEntry instead of IsPvpZone which will check for the zoneID's... it's not perfect tho and will require some more checks for wintergrasp & tol barad (although maybe we can let scripts handle that).

File: src/server/scripts/Spells/spell_generic.cpp
Comment (joschiwald) [inline src/server/scripts/Spells/spell_generic.cpp]: what does SPELL_AURA_119 do?

File: src/server/scripts/Spells/spell_generic.cpp
Comment (soulfrost) [inline src/server/scripts/Spells/spell_generic.cpp]: parent spell 134732 shouldnt be applyed on player?

File: src/server/scripts/Spells/spell_generic.cpp
Comment (Golrag) [inline src/server/scripts/Spells/spell_generic.cpp]: I don't know but seeing how only this spell and 211128 Black Rook Rumble has it, it might control pvp talents :p

File: src/server/scripts/Spells/spell_generic.cpp
Comment (Golrag) [inline src/server/scripts/Spells/spell_generic.cpp]: I've not seen this spell in my sniffs

File: src/server/scripts/Spells/spell_generic.cpp
Comment (Golrag) [inline src/server/scripts/Spells/spell_generic.cpp]: 211128 does enable pvp talents so I'm working on making spell_aura_119 enable/disable pvp talents

**Comment by Golrag** [comment]
For pvp combat it seems like SMSG_AURA_UPDATE is kinda being spammed with the 20sec duration for 134735 PvP Rules Enabled. (Without it being cast when re-applying). Got this from sniffs with the World Quest black rook rumble. So in the warrior arena it's applied without duration and here it seems to be applied with it.

**Comment by Golrag** [comment]
I removed some of the AreaID checks (they were only enabled during World Quests). I would like to remove those warden tower areas as well, still researching if world quest system is involved for those.
Also enabled pvp talents in battlefields even when they are not started (checked this today with wintergrasp and tol barad).

File: src/server/game/DataStores/DB2Structure.h
Comment (Aokromes) [inline src/server/game/DataStores/DB2Structure.h]: i wonder if there is a generic flag to check this.

Comment (Golrag) [inline src/server/game/DataStores/DB2Structure.h]: For warden tower zones:
AREA_FLAG_UNK9 | AREA_FLAG_ALLOW_DUELS
+- 5099 areas have this flag | +- 4472 areas have this flag

For tol barad:
AREA_FLAG_OUTLAND2 | AREA_FLAG_ALLOW_DUELS

I think some flags have bad naming but I don't have the knowledge to research this :/

Comment (Golrag) [inline src/server/game/DataStores/DB2Structure.h]: QuestInfoID 113 causes it I think.
```
SELECT ID, QuestType, LogTitle FROM quest_template WHERE QuestInfoID=113;
+-------+-----------+---------------------------------------------+
| ID    | QuestType | LogTitle                                    |
+-------+-----------+---------------------------------------------+
| 37295 |         3 | Kill Rylai Crestfall                        |
| 37297 |         3 | Kill Jeron Emberfall                        |
| 37299 |         3 | Bonus Objective: Kill Stormshield Priests   |
| 37475 |         3 | Molten Quarry                               |
| 37476 |         3 | Ashmaul Burial Grounds                      |
| 37477 |         3 | Brute's Rise                                |
| 37478 |         3 | Amphitheater of Annihilation                |
| 37479 |         2 | Bonus Objective: Kill Alliance Leaders      |
| 37480 |         2 | Bonus Objective: Kill Alliance Leaders      |
| 43601 |         3 | Warden Tower Assault: Whisperwind's Citadel |
| 43600 |         3 | Warden Tower Assault: Whisperwind's Citadel |
| 43599 |         3 | Warden Tower Assault: Blackhawk's Bulwark   |
| 43598 |         3 | Warden Tower Assault: Blackhawk's Bulwark   |
| 43303 |         2 | Time to Rumble!                             |
| 43248 |         3 | Warden Tower Assault: Darkfollow's Spire    |
| 43247 |         3 | Warden Tower Assault: Darkfollow's Spire    |
| 43183 |         3 | Warden Tower Assault: Starstalker's Point   |
| 42070 |         3 | Warden Tower Assault: Starstalker's Point   |
| 42023 |         3 | Black Rook Rumble                           |
| 41421 |         3 | Warden Tower Assault: Nightwatcher's Perch  |
| 41420 |         3 | Warden Tower Assault: Nightwatcher's Perch  |
| 41257 |         3 | Warden Tower Assault: Cordana's Apex        |
| 41227 |         3 | Warden Tower Assault: Cordana's Apex        |
| 41013 |         3 | Darkbrul Arena                              |
| 42025 |         3 | Bareback Brawl                              |
| 41896 |         3 | Operation Murloc Freedom                    |
| 45740 |         3 | Silithyst                                   |
| 45657 |         3 | Silithyst                                   |
+-------+-----------+---------------------------------------------+
```

Will remove it all those warden tower areas

**Comment by HannibalRoG** [comment]
@Shauren need your review here.

**Comment by soulfrost** [comment]
no all checks here, lot of missing GameUI errors, which checked by client and can be hacked

**Comment by Golrag** [comment]
Could you give me some examples ? :)

**Comment by soulfrost** [comment]
ERR_TALENT_FAILED_CANT_REMOVE_TALENT                    = 810,
    ERR_TALENT_FAILED_REST_AREA                             = 979,

- related to GameUiError packet ( id can be a bit changed )

**Comment by Golrag** [comment]
I added SMSG_LEARN_PVP_TALENTS_FAILED but the reason it uses is still the same as the one we use for normal talents. Not sure if that's an answer to what you meant tho :s

File: src/server/game/DataStores/DB2Stores.cpp
Comment (Golrag) [inline src/server/game/DataStores/DB2Stores.cpp]: Small space I have to remove :p

File: src/server/game/Entities/Player/Player.cpp
Comment (Golrag) [inline src/server/game/Entities/Player/Player.cpp]: Should only check for pvp combat. you can attack an enemy player, then engage an npc and if you wait long enough your pvp talents will be disabled while you're still in combat with npc. Is there a way to check this ?
Also I have seen this buff on npc's so I will have to move it to Unit and check if the behaviour is the same

Comment (Golrag) [inline src/server/game/Entities/Player/Player.cpp]: @Treeston does your PR #19930 provide anything for pvp combat ? :)
I see that there is ```bool HasPvPCombat()``` does it do what I need here ?

Comment (Treeston) [inline src/server/game/Entities/Player/Player.cpp]: Yes, it should do exactly what you want here.

Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: Or, you know, any unit pvp flagged?

Comment (Golrag) [inline src/server/game/Entities/Player/Player.cpp]: I tried something with ```m_ThreatManager.getThreatList();``` in the meantime but it returned an empty list while in combat D:

Comment (Treeston) [inline src/server/game/Entities/Player/Player.cpp]: Is there a `m_combatTimer` equivalent for master? IIRC that's only nonzero if in PvP combat.

**Comment by Golrag** [comment]
I noticed that when you have some pvp talents and you downgrade your character (like going from lvl 110 to lvl 109) your pvp talents will still be learned when you get in pvp, is this we should handle ? If so what would be the best way ?

**Comment by Shauren** [comment]
You mean with a GM command?

**Comment by Golrag** [comment]
yes

**Comment by Treeston** [comment]
The best (and only) way would be to figure out _why_ this happens.

**Comment by Golrag** [comment]
Missing check in ```Player::GiveLevel```. I wonder what will happen once they raise level cap, will the minimum level for pvp talents change ?

**Comment by Zydrax** [comment]
Is not working to learn Honorable Medallion

**Comment by Golrag** [comment]
@Zydrax could you provide more info ? It's working for me :s (not the achievement tho but I think that's a little bit out of scope)

**Comment by Zydrax** [comment]
Sory is work , but only whit command , is not work learn from talent

File: src/server/game/Spells/Auras/SpellAuraEffects.cpp
Comment (joschiwald) [inline src/server/game/Spells/Auras/SpellAuraEffects.cpp]: if (Player* target = auraApp->GetTarget()->ToPlayer())

**Comment by Treeston** [comment]
By the way: don't merge master into your branch, rebase your branch on master instead.

**Comment by Golrag** [comment]
@Shauren have your requested changes being addressed correctly or is there something I missed ? :) Would be great if the "Feedback-FixOutdatedMissingWIP" could be removed :p

**Comment by Golrag** [comment]
Anyone knows if something changed with 7.3.0 / 7.3.2 ? :)

**Comment by Shauren** [comment]
None of PvpTalent db2s changed

**Comment by DoctorKraft** [comment]
@Golrag Any news with this PR, something missing?

**Comment by tkrokli** [comment]
Except for commit https://github.com/TrinityCore/TrinityCore/commit/769db08a67db67e2dca7910c23277fa94d98fe7f, probably just busy IRL.

**Comment by Golrag** [comment]
Not that I am aware of. I think I addressed the requested changes correctly. Didn't notice any bugs while testing on 7.2.5. I also did some quick testing on 7.3.2 without issues.
I also got mentions of other people merging this in their own project so I hope they report any issues they find

**Comment by tkrokli** [comment]
OK, so we are mostly waiting for Shauren to approve the changes.

**Comment by Golrag** [comment]
Okay found a bug.
Start duel, with someone of your faction
Other faction player enter combat with you
End duel (while other faction player still in combat with you)
Run away
Wait for combat to end and wait 20sec. Pvp talents are still enabled on character who was in the duel

**Comment by Golrag** [comment]
They are going to change the way pvp talents work in bfa, which will be in like 5 months (prepatch maybe) is there anything left to do for me here ? :p

**Comment by Shauren** [comment]
No, I will merge this soon (once phasing is done), you don't even need to update structures to 7.3.5

**Comment by Shauren** [comment]
Small update - will load World_PVP_Area.db2 for ActivatesPvpTalents()

**Review by Shauren — APPROVED** [review]

**Comment by Golrag** [comment]
@Shauren You sure about battlefields only activating them when they are in war time ? On a pvp realm it seems to be activated when you enter them event if they are not in war time, might be different on a normal realm

**Comment by Shauren** [comment]
Tested on normal realm, they are enabled too

**Comment by Golrag** [comment]
Thanks for putting your time in it to merge it :) do you mind giving me a small recap of your changes so I could learn of it ?
If it's too much time I could check it out with my local changes still applied and make a diff of it

**Comment by Shauren** [comment]
- ASSERTs for PvpTalentUnlockEntry loading
- added class arg to DB2Manager::GetPvpTalentsByPosition
- obviously structure update for 7.3.5
- added Role check for learning talents to prevent cheating (there are talents shared by multiple classes but healers have something else in that slot)
- Player::IsAreaThatActivatesPvpTalents

**Pull request merged**
